packages/hash/frontend/src/lib/entities.ts

```diff
@@ -1,127 +1,460 @@
+      `sourceEntityId has already been re-written as a stringified object: ${sourceEntityId}`,
+    );
+  }
+
+  return {
+    aggregationId: rewriteLinkedAggregationIdentifier({
+      accountId: sourceAccountId,
+      aggregationId,
+    }),
+    sourceEntityId: rewriteEntityIdentifier({
+      accountId: sourceAccountId,
+      entityId: sourceEntityId,
+    }),
+    path,
+    operation: operation as BpLinkedAggregation["operation"],
+    results:
+      "results" in linkedAggregation && linkedAggregation.results
+        ? convertApiEntitiesToBpEntities(linkedAggregation.results)
+        : undefined,
+  };
+}
+
+/**
+ * Converts an entity type from its GraphQL API representation to its Block Protocol representation:
+ * 1. Only provide 'entityTypeId' at the top level
+ * 2. Provide the schema under 'schema', not 'properties'
  *
- * Deletes existing root fields unless they relate to the entity's id or type.
- * To preserve all existing root fields, pass 'false' as the second argument.
- * @param entity The entity to clone
- * @param preserveExtraMetadata Whether to keep non-id and non-type metadata.
- * Defaults to FALSE: extra metadata (e.g. accountId, visibility) will be deleted.
- * @returns the entity tree clone, transformed.
+ * N.B. this intentionally does not re-write 'entityTypeId' to include accountId, since types are not sharded,
+ * and the 'entityTypeId' is sufficient to identify entity types when calling the HASH API.
  */
-export const cloneEntityTreeWithPropertiesMovedUp = (
-  entity: Partial<UnknownEntity>,
-  preserveExtraMetadata: boolean = false,
-) => {
-  const clonedTree = cloneEntityWithPropertiesAtRoot(
-    entity,
-    preserveExtraMetadata,
-  );
-
-  const propertiesToCheck = Object.values(clonedTree).filter(
-    isParsedJsonObjectOrArray,
-  );
-
-  while (propertiesToCheck.length > 0) {
-    const property = propertiesToCheck.pop()!;
-
-    if (Array.isArray(property)) {
-      propertiesToCheck.push(...property.filter(isParsedJsonObjectOrArray));
-      continue;
-    }
-
-    if (
-      !property.entityId ||
-      !property.properties ||
-      !isParsedJsonObject(property.properties)
-    ) {
-      // This is a non-entity object - it might have entities deeper in its tree
-      propertiesToCheck.push(
-        ...Object.values(property).filter(isParsedJsonObjectOrArray),
-      );
-      continue;
-    }
-
-    propertiesToCheck.push(
-      ...Object.values(property.properties).filter(isParsedJsonObjectOrArray),
-    );
-
-    destructivelyMoveEntityPropertiesToRoot(property, preserveExtraMetadata);
-  }
-  return clonedTree;
+export const convertApiEntityTypeToBpEntityType = ({
+  entityId,
+  entityTypeId,
+  properties,
+}: Pick<
+  ApiEntityType,
+  "entityId" | "entityTypeId" | "properties"
+>): BpEntityType => {
+  if (entityTypeId.includes("{")) {
+    throw new Error(
+      `entityTypeId has already been re-written as a stringified object: ${entityTypeId}`,
+    );
+  }
+  return {
```

Comment:
**Blocking:** similar to my previous comment, why are we checking whether `entityTypeId` is a stringified object if we aren't stringifying it anywhere? Perhaps this should be checking whether `entityId` is stringified?